Add an early callback

kaduk opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated
- [x] CLA is signed

##### Description of change

Implement an early callback similar to BoringSSL's select_certificate_cb, allowing for SNI processing and server certificate selection prior to session resumption and version negotiation.

The chosen API (with opaque structure and accessors, rather than public structure) differs from BoringSSL's, in order to remain consistent with most other OpenSSL APIs, and retain the trend of opaque data structures that facilitate ABI stability.  This also should make it easier to backport the API to 1.1.0 if desired.

The WIP label will remain until tests are added, but other interesting questions include:

* should the raw clienthello bytes be exposed at all?

* Is now the time to implement a TLS extension type to index lookup, to avoid having to loop over the preprocessed extensions?

* What level of support should SSLv2 clienthellos get from this API?  (I am not much of an expert on the SSLv2 wire protocol, so the current support is likely lacking.)

* Should we expose the alert to send to the callback, or even allow the callback to terminate the connection at all?


cc @mattcaswell

kaduk commented:
I guess also:

* Should the callback be able to mark an extension as "processed"

I also got one request out of band to more clearly document the callback's responsibility to handle SSLv2 ClientHellos more clearly.

davidben commented:
I notice this callback is blocking only and does not have a way to pause the state machine like `cert_cb` does. You could imagine situations where looking up the certificate or other configuration needs to be asynchronous.

For instance, Node.js's `SNICallback` [here](https://nodejs.org/api/tls.html#tls_tls_createserver_options_secureconnectionlistener) is asynchronous. Looks like they currently query `SSL_get_servername` from `cert_cb` (which does support this), but this hits the session resumption ordering issues alluded to in your man page. @indutny may be able to speak to Node's needs more.


> Is now the time to implement a TLS extension type to index lookup, to avoid having to loop over the preprocessed extensions?

I too twitch when I see such O(N) calls. :-) But I don't think it actually matters for this. There's no reason to ever write code like:

    for (uint16_t ext = 0; ext < 0xffff; ext++) {
      SSL_EARLY_get0_ext(ssl, ectx, ext, ...);
      ...
    }

The callback would ultimately only query the three or five specific extension types it cares about, so it wouldn't go quadratic.

> What level of support should SSLv2 clienthellos get from this API? (I am not much of an expert on the SSLv2 wire protocol, so the current support is likely lacking.)

One detail to note is the cipher list format is different (2-byte vs 3-byte cipher values), so you'll need to decide whether you convert internally or force the caller to handle both cases. (Our callback acts after V2ClientHellos are converted to SSLv3/TLS ClientHellos, so the callback never sees V2ClientHellos. Though I see OpenSSL 1.1.x uses a different strategy for V2ClientHello now.)

The random is also different, but less likely to trip callers up (or even be queried): in SSLv3/TLS it has fixed size, while in a V2ClientHello, the size is variable.

> Should the callback be able to mark an extension as "processed"

Why? What would that do?

kaduk commented:
> > Should the callback be able to mark an extension as "processed"

> Why? What would that do?

The idea was to prevent the builtin processing of that extension from taking place, though I realize now that I may be misremembering what that field in the structure is used for.

indutny commented:
I'd like to sincerely ask you to consider doing asynchronous version of this. There is just no functionality that can substitute this, and both Node.js and [bud](https://github.com/indutny/bud) need this to some extent.

davidben commented:
> The idea was to prevent the builtin processing of that extension from taking place, though I realize now that I may be misremembering what that field in the structure is used for.

That sounds undesirable. Extensions aren't independent sub-protocols; they're how we add new ClientHello fields. These may or may not interact. (Of course, we usually make them not interact because it's simpler that way, but not always. The crypto negotiation extensions used in TLS 1.3 are fairly coupled together.)

If you have two extensions which do interact, having a low-level "prevent built-in processing" knob means you expose implementation details. Perhaps you happen to implement it such that feature X is keyed on extension A rather than B.

It's also not clear to me why you'd want to do that anyway. I can imagine "please disable high-level feature X", but that should be done by having an API to disable said feature, not by shutting off low-level extensions via this callback.

kaduk commented:
> > prevent the builtin processing of that extension

> That sounds undesirable.

Yeah, I was pretty sure I didn't want it, but figured I could mention it explicitly so we consciously decide to not do it.

richsalz commented:
We don't want to prevent the builtin processing.
If someone needs to do that, hack the source.

kaduk commented:
> One detail to note is the cipher list format is different (2-byte vs 3-byte cipher values)

#2280 seems relevant

tmshort commented:
Does anything even send SSLv2-compatible ClientHellos any more?

tmshort commented on ssl/statem/extensions.c:
```diff
@@ -357,11 +357,12 @@ static int extension_is_relevant(SSL *s, unsigned int extctx,
  * extensions that we know about. We ignore others.
  */
-int tls_collect_extensions(SSL *s, PACKET *packet, unsigned int context,
+size_t tls_collect_extensions(SSL *s, PACKET *packet, unsigned int context,
```
If this is a number of extension, not sure it needs to be a size_t; the original int is likely sufficient, given the maximum size of a ClientHello.

kaduk replied:
This has theh type match with clienthello.pre_proc_exts_len, which is what we're going to assign it to, but I agree that int is going to be big enough to hold the values that would be seen.

mattcaswell replied:
I prefer not to overload the return type to mean different things, i.e. I'd prefer a simple boolean (int) return type to indicate success or failure, with a separate `size_t *` to set the length. Incidentally the length should *definitely* be a size_t and not an int in order to match the types. I have seen far too many subtle bugs introduced by lengths swapping backwards and forwards between int and size_t. They should be consistent throughout. I don't want to have to redo all the size_t work I did in libssl just recently.

kaduk replied:
Okay, I'll change to a separate output parameter.

tmshort commented:
All session handling is done via the initial_ctx/session_ctx (ugh, two names for the same thing), so, if you want to change how resumed sessions are handled, an API to change the initial_ctx/session_ctx needs to be proposed. Right now, sessions from different virtual servers may share the same resumption database (and ticket crypto). This may be desirable if those virtual servers are on the same machine, but not so much if certificate authentication is being used.

tmshort commented on doc/man3/SSL_CTX_set_early_cb.pod:
```diff
@@ -0,0 +1,94 @@
+                                 const unsigned char *clienthello, size_t ch_len,
+                                 int *al, void *arg);
+ void SSL_CTX_set_early_cb(SSL_CTX *c, ssl_early_cb_fn cb, void *arg);
+ int SSL_EARLY_isv2(SSL *s, SSL_EARLY *ectx);
```
Typically, SSL_EARLY would be the first type in a function named as such, so perhaps not include the SSL structure, and "hide" it inside the SSL_EARLY? Maybe not? But the name might need to be changed to make it consistent with other APIs.

kaduk replied:
I thought about that, but the temptation to make SSL_EARLY just be reusing an existing type (CLIENTHELLO_MSG) was too strong, at least for the first crack.  The callback is almost certainly going to need the SSL object itself, though, so it seems reasonable to have it be a top-level API parameter.  I guess we could go the other way and attach the CLIENTHELLO_MSG to the SSL before invoking the callback, though.

mattcaswell replied:
If the function name is `SSL_EARLY_*` then the `SSL_EARLY` argument should be the first IMO

davidben commented:
> Does anything even send SSLv2-compatible ClientHellos any more?

We recently added code to get some data on our end. I would love to lose the V2ClientHello code, but I'm not optimistic.

indutny commented:
Just to re-state my inquiry, it would be great if this could work asynchronously.

kaduk commented:
> Just to re-state my inquiry, it would be great if this could work asynchronously.

Acknowledged.  Last week was a bit busy around here...
(Also it looks like it will need a little bit of re-plumbing compared to cert_cb.)

kaduk commented:
> if you want to change how resumed sessions are handled, an API to change the initial_ctx/session_ctx needs to be proposed

I am a little reluctant to add that into this changeset (which is going to be getting bigger once I add the ability for the callback to be async and tests.  It seems like it could potentially be done as a follow-up change, if there is even consensus from the team that it should be done.

tmshort commented:
It's really more of a question, is modifying session resumption part of the intended functionality? If yes, then an API will need to be introduced.

kaduk commented:
No async yet (still considering options), but this adds tests and exposes a helper to convert raw cipherlist octets into SSL_CIPHERs, that would probably be useful enough to applications to be worth exposing.
Removing the WIP label to seek more review, though I expect there will be several things that need changing (not least, to be compatible with the various no-foo options; hopefully travis will help with that).

mattcaswell commented on doc/man3/SSL_CTX_set_early_cb.pod:
```diff
@@ -0,0 +1,94 @@
+
+SSL_CTX_set_early_cb() sets the callback function, which is automatically
+called during the early stages of ClientHello processing on the server.
+The argument supplied whens setting the callback is passed back to the
```
typo: whens

mattcaswell commented on doc/man3/SSL_get_ciphers.pod:
```diff
@@ -63,10 +73,19 @@ free the return value itself.
+The stack returned by SSL_bytes_to_cipher_list() should be freed using
+sk_SSL_CIPHER_free().
+
 =head1 RETURN VALUES
```
Presumably `SSL_bytes_to_cipher_list()` returns NULL on failure (e.g. a bad length)

kaduk replied:
Actually, ended up changing the return convention to use an output pointer, since we want to return SCSVs separately so that the state machine can act on them.

mattcaswell commented on doc/man3/SSL_get_ciphers.pod:
```diff
@@ -63,10 +73,19 @@ free the return value itself.
+
+The implementation of SSL_bytes_to_cipher_list() mutates state in the
+supplied SSL object B<s>; SSL_bytes_to_cipher_list() should not be called
+on a server SSL object after that server has processed the received ClientHello.
```
Yuck. That sounds like a really bad idea. Surely we can fix this?

mattcaswell commented on include/openssl/ssl.h:
```diff
@@ -1558,6 +1558,27 @@ __owur char *SSL_get_srp_username(SSL *s);
+/*
+ * Early callback and helpers.
+ */
+typedef struct clienthello_msg_st SSL_EARLY;
```
Hmmm.....we should perhaps change the name of clienthello_msg_st. That's going to get really confusing.

kaduk replied:
Any suggestions?  Alternately we can embed clienthello_msg* and SSL* into a new ssl_early_st, as mentioned above.  Might be worth the extra type...

mattcaswell commented on include/openssl/ssl.h:
```diff
@@ -1558,6 +1558,27 @@ __owur char *SSL_get_srp_username(SSL *s);
+typedef int (*ssl_early_cb_fn) (SSL *s, SSL_EARLY *ectx,
+                                const unsigned char *clienthello, size_t ch_len,
+                                int *al, void *arg);
+void SSL_CTX_set_early_cb(SSL_CTX *c, ssl_early_cb_fn cb, void *arg);
```
Is there no need for a `SSL_set_early_cb()`?

richsalz replied:
uppercase the SSL in the cb typedef.

kaduk replied:
Changed to using the majuscule in the typedef name.
There ... probably is not a need for a SSL_set_early_cb(), though I'm happy to put one in if you want.  Basically, the early callback itself is the first time you will get any data that could possibly make something unique per SSL derived from the same parent SSL_CTX, AIUI.  So, there isn't a "legitimate" need for it, just possibly one of convenience in some application coding styles.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -4343,6 +4343,80 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
 }
 
+void
+SSL_CTX_set_early_cb(SSL_CTX *c, ssl_early_cb_fn cb, void *arg)
```
Please put the return  type on the same line as the rest of the declaration. Same for other funcs below.

kaduk replied:
Gah, this is my other projects' style creeping in; I always forget that OpenSSL doesn't like this way.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -4343,6 +4343,80 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+{
+    if (out != NULL)
+        *out = ectx->random;
+    return ectx->isv2 ? 0 : SSL3_RANDOM_SIZE;
```
Why is it necessary to distinguish whether it is v2 here or not? If it is v2 then we use the variable length challenge as the random. If it is longer than SSL3_RANDOM_SIZE then we just use the first SSL3_RANDOM_SIZE bytes. If it is less than SSL3_RANDOM_SIZE then we pad to the left with 0s until it is SSL3_RANDOM_SIZE in length. Either way the resulting length is SSL3_RANDOM_SIZE in all cases.

kaduk replied:
To be sure that someone who actually is familiar with the v2 protocol (i.e., not me) looks at this logic :)
I'm happy to take your word for it; it didn't really seem worth putting the time in to understand the v2-parsing code given that it's (hopefully) on its way out.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -4343,6 +4343,80 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+}
+
+unsigned int
+SSL_EARLY_get0_version(SSL *s, SSL_EARLY *ectx)
```
Should we call this legacy_version?

kaduk replied:
Yes.

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -4343,6 +4343,80 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+    RAW_EXTENSION *r;
+
+    /* XXX potential accidental quadratic behavior */
+    for (i = 0; i < ectx->pre_proc_exts_len; ++i) {
```
`pre_proc_exts_len` is the number of built in extensions plus the number of custom extensions that are defined. It is *not* the number of extensions that were sent. If we don't recognise them then they don't end up in pre_proc_exts. So this is bounded by the server config not by how many extensions we get sent. That said and index -> type lookup might not be a bad idea....or maybe expose the index????

kaduk replied:
I'll remove the XXX comment, given this and davidben's comment.

> That said and index -> type lookup might not be a bad idea....or maybe expose the index????

Do you mean type->index lookup?  I had thought about doing that at some point, but demurred given that (1) there wasn't one already, possibly intentionally, and (2) I didn't want to think about how it would interact with custom extensions right then.

mattcaswell commented on test/handshake_helper.c:
```diff
@@ -144,6 +188,36 @@ static int servername_reject_cb(SSL *s, int *ad, void *arg)
+                          size_t ch_len, int *al, void *arg)
+{
+    int ret = early_select_server_ctx(s, ectx, arg, 0);
+    SSL_set_options(s, SSL_OP_NO_TLSv1_2);
```
Use `SSL_set_max_proto_version()`

kaduk replied:
Good catch, thanks.

kaduk commented:
The latest push is just the minor fixes and making SSL_bytes_to_cipher_list() "pure" -- async is still WIP.
But, it seems that the way to do that is to move the last ~half of tls_process_client_hello() (the early callback and later) into a separate function and do all that stuff from the "post_work" portion of the state machine.
(Matt notes that this will require adding a new WORK_MORE_X state.)

kaduk commented:
And, landed the async functionality.
The sslv2-in-sslv2 test currently fails since we now send an alert and the test is expecting no alert.  (This is a side effect of the refactoring, as there's not an easy no-alert error path from the subroutine moved from the "processing" to the "post-processing" stage.  Should we just adopt the test to the new behavior?
I also took out the raw clienthello bytes as direct arguments to the early callback, as it got in the way of the refactoring needed for the async behavior, and I was not entirely keen on it in the first place.  (This does not preclude us from later adding an accessor to get the raw bytes, of course, it just helps motivate people to use the accessors instead of rolling their own parser.)
@mattcaswell please take another look

kaduk commented on test/handshake_helper.c:
```diff
@@ -144,6 +188,33 @@ static int servername_reject_cb(SSL *s, int *ad, void *arg)
+{
+    int ret = early_select_server_ctx(s, ectx, arg, 0);
+    SSL_set_max_proto_version(s, TLS1_1_VERSION);
+    if (!ret)
```
I guess we could add some calls to the SSL_EARLY_* accessors in here to stop coveralls from complaining.
(And please confirm that I should make the change to putting the SSL_EARLY object first, or make a new type for it to avoid the need to pass  the SSL as well, or ...)

kaduk commented on ssl/ssl_lib.c:
```diff
@@ -4457,3 +4525,160 @@ int ssl_log_secret(SSL *ssl,
+{
+    int n;
+
+    s->s3->send_connection_binding = 0;
```
This arguably ought to be part of the SCSV processing, but I didn't want to change the order of processing more than needed, at first.

kaduk commented on doc/man3/SSL_CTX_set_early_cb.pod:
```diff
@@ -0,0 +1,92 @@
+
+It is recommended that application callbacks make use of the provided
+SSL_EARLY_* APIs instead of attempting to decode and parse the raw
+ClientHello bytes directly.
```
Hmm, this needs to go if we decide to keep the commits that remove this functionality.

kaduk commented on ssl/statem/statem_srvr.c:
```diff
@@ -1191,13 +1178,13 @@ MSG_PROCESS_RETURN tls_process_client_hello(SSL *s, PACKET *pkt)
     /*
-     * First, parse the raw ClientHello data into the CLIENTHELLO_MSG structure.
+     * First, save a pointer to the raw ClientHello data for the early callback,
+     * and parse it into the CLIENTHELLO_MSG structure.
```
Also need to revert this comment back, if we keep the commits that remove access to the raw ClientHello octets.

kaduk commented on ssl/statem/statem_srvr.c:
```diff
@@ -1704,66 +1812,22 @@ MSG_PROCESS_RETURN tls_process_client_hello(SSL *s, PACKET *pkt)
+    return 1;
  f_err:
-    ssl3_send_alert(s, SSL3_AL_FATAL, al);
+    /* The caller will send the alert for us. */
```
This is related to the test failure I mentioned in a previous comment.

mattcaswell commented on test/ssl-tests/05-sni.conf.in:
```diff
@@ -109,4 +109,55 @@ our @tests = (
+            },
+        },
+        test   => {
+            "ExpectedProtocol" => "TLSv1.1",
```
This test fails with the `no-tls1_1` config option

mattcaswell replied:
Take a look at 20-cert-select.conf.in for an example of how to do this type of thing. Some of the other ".in" files do similar things.

kaduk replied:
Ah, thanks.  I got there ... eventually ... I should probably call it quits for the night if I'm making this many brainos.

kaduk commented:
Decided to not push -f for this round of changes; please let me know if/when I should rebase/squash.
(Not everything that could be squashed is explicitly noted as such in the commit message.)

kaduk commented:
(Had to put the clienthellomsg_st tag back, as I misremembered how things were implemented; sorry for the force-push, though it only affected the last commit or two.)

kaduk commented on doc/man3/SSL_CTX_set_early_cb.pod:
```diff
@@ -0,0 +1,84 @@
+ size_t SSL_early_get0_session_id(SSL *s, const unsigned char **out);
+ size_t SSL_early_get0_ciphers(SSL *s, const unsigned char **out);
+ size_t SSL_early_get0_compression_methods(SSL *s, const unsigned char **out);
+ int SSL_early_get0_ext(SSL *s, int type, const unsigned char **out,
```
done

kaduk commented on doc/man3/SSL_CTX_set_early_cb.pod:
```diff
@@ -0,0 +1,84 @@
+The argument supplied when setting the callback is passed back to the
+callback at runtime.  Error returns from the callback will cause the
+connection to terminate, and callbacks returning failure should indicate
+what alert value is to be sent in the B<al> parameter.
```
It turns out that a bit more plumbing was still needed to bring the async part of the early callback into feature parity with similar things such as the cert callback.  I've added more documentation for that case along with the extra code.

kaduk commented on doc/man3/SSL_CTX_set_early_cb.pod:
```diff
@@ -0,0 +1,84 @@
+
+SSL_early_get0_random(), SSL_early_get0_session_id(), SSL_early_get0_ciphers(),
+and SSL_early_get0_compression_methods() return the length of the corresponding
+ClientHello fields.
```
(see below)

kaduk commented on doc/man3/SSL_CTX_set_early_cb.pod:
```diff
@@ -0,0 +1,84 @@
+
+L<ssl(7)>, L<SSL_CTX_set_tlsext_servername_callback(3)>
+
+=head1 COPYRIGHT
```
done

kaduk commented on doc/man3/SSL_get_ciphers.pod:
```diff
@@ -41,6 +44,14 @@ SSL_get_client_ciphers() returns the stack of available SSL_CIPHERs matching the
+SSLv3/TLS wire format), and parses the cipher suites supported by the library
+into the returned stacks of SSL_CIPHER objects sk and Signalling Cipher-Suite
+Values scsvs.  Unsupported cipher suites are ignored.  Returns 1 on success
+and 0 on failure.
```
It looks like it was a remnant of a previous version.  Since we're going for making this public API "pure", there is not really a need to indicate back what alert would be sent, so I think it should just be removed.

kaduk commented on include/openssl/ssl.h:
```diff
@@ -2143,7 +2161,7 @@ int ERR_load_SSL_strings(void);
 # define SSL_F_SSL_ADD_SERVERHELLO_USE_SRTP_EXT           308
 # define SSL_F_SSL_BAD_METHOD                             160
 # define SSL_F_SSL_BUILD_CERT_CHAIN                       332
-# define SSL_F_SSL_BYTES_TO_CIPHER_LIST                   161
```
it was in 1.1.0; restored.

kaduk commented on ssl/ssl_lib.c:
```diff
@@ -4360,6 +4361,72 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+
+unsigned int SSL_early_get0_legacy_version(SSL *s)
+{
+    return s->clienthello->legacy_version;
```
:)
I put the check into SSL_early_get0_ex() largely to prompt this question, namely, is the lack of s->clienthello an internal error to be bubbled up to the application, or an application error, for calling an SSL_early_* API from an invalid context?  I am inclined to go with the latter, since the handshake ends up in an error state if we fail to set s->clienthello to a valid value (okay, I had to go add the NULL check just now, but it's clear it should have always been there) and it should stay there through any valid codepaths that end up in the early callback being called.
That is, I favor removing the check from SSL_early_get0_ext() and asserting that the SSL_early() functions cannot fail [in correct usage].

mattcaswell replied:
Generally our pattern seems to be we don't check the primary object (i.e. SSL) for NULL, but we do check other things. It would only be NULL through programmer error (calling when not in an early callback), but - still, we probably shouldn't crash.

kaduk replied:
Okay, I'll make them return 0 as the least-bad option (unless you want proper success/failure return codes and output parameters, but I don't get that impression).

kaduk commented on ssl/ssl_lib.c:
```diff
@@ -4457,3 +4524,159 @@ int ssl_log_secret(SSL *ssl,
+
+int SSL_bytes_to_cipher_list(SSL *s, const unsigned char *bytes, size_t len,
+                             int isv2format, STACK_OF(SSL_CIPHER) **sk,
+                             STACK_OF(SSL_CIPHER) **scsvs, int *al)
```
yup; it's gone now.

kaduk commented on ssl/ssl_locl.h:
```diff
@@ -1127,6 +1131,9 @@ struct ssl_st {
     } ext;
 
+    /* Parsed form of the ClientHello, kept around across early_cb calls. */
+    struct clienthello_msg_st *clienthello;
```
IIRC, the typedef is later in the file, so we'd have to reorder things in order to use it here.

mattcaswell replied:
So lets move the typedef earlier :-)

kaduk replied:
done (also RAW_EXTENSION that it depends on)

kaduk commented on ssl/statem/statem_srvr.c:
```diff
@@ -1340,105 +1326,194 @@ MSG_PROCESS_RETURN tls_process_client_hello(SSL *s, PACKET *pkt)
+        if (code == 0)
+            goto err;
+        if (code < 0)
+            return code;
```
An interesting question.  It turns out that there was more plumbing needed to get the desired async functionality; I think I got the rest of it, but of course we won't be sure until it's tested.

kaduk commented:
Okay, plumbed in a test of the stop-and-restart functionality into sslapitest.
I think this is the last of the changes @mattcaswell requested.  Let me know if you want me to rebase and/or squash before you look again.

FdaSilvaYY commented on test/sslapitest.c:
```diff
@@ -477,6 +477,83 @@ static int test_keylog_no_master_key(void) {
+{
+    int *ctr = arg;
+    const unsigned char *p;
+    /* We only configure two ciphres, but the SCSV is added automatically. */
```
typo on `ciphres`

FdaSilvaYY commented on test/sslapitest.c:
```diff
@@ -477,6 +477,83 @@ static int test_keylog_no_master_key(void) {
+    int *ctr = arg;
+    const unsigned char *p;
+    /* We only configure two ciphres, but the SCSV is added automatically. */
+    const unsigned char expected_ciphers[] = {0x00, 0x9d, 0xc0,
```
static const ?

kaduk replied:
This is in function scope, so I don't feel a particular need for static.

kaduk commented:
This needs a rebase to catch up with some changes on master, FWIW (I have done so locally already but held off on pushing).  And if @mattcaswell confirms that it's on his radar I promise to not pester him about it for a while (a week?).

mattcaswell commented on doc/man3/SSL_get_ciphers.pod:
```diff
@@ -14,6 +14,9 @@ SSL_get_ciphers, SSL_CTX_get_ciphers, SSL_get_cipher_list
  STACK_OF(SSL_CIPHER) *SSL_get_client_ciphers(const SSL *ssl);
+ int SSL_bytes_to_cipher_list(SSL *s, const unsigned char *bytes, size_t len,
+                              int isv2format, STACK_OF(SSL_CIPHER) **sk,
+                              STACK_OF(SSL_CIPHER) **scsvs)
```
nit: missing ;

kaduk replied:
done

mattcaswell commented on include/openssl/ssl.h:
```diff
@@ -772,6 +773,7 @@ __owur int SSL_extension_supported(unsigned int ext_type);
 # define SSL_want_x509_lookup(s) (SSL_want(s) == SSL_X509_LOOKUP)
 # define SSL_want_async(s)       (SSL_want(s) == SSL_ASYNC_PAUSED)
 # define SSL_want_async_job(s)   (SSL_want(s) == SSL_ASYNC_NO_JOBS)
+# define SSL_want_early(s)       (SSL_want(s) == SSL_EARLY_WORK)
```
Please update the SSL_want and SSL_get_error pod pages

kaduk replied:
done

mattcaswell commented on ssl/ssl_lib.c:
```diff
@@ -3076,6 +3077,9 @@ int SSL_get_error(const SSL *s, int i)
     if (SSL_want_async_job(s)) {
         return SSL_ERROR_WANT_ASYNC_JOB;
     }
+    if (SSL_want_early(s)) {
```
No braces needed here...might as well fix the other ones while you're at it ;-)

kaduk replied:
Yeah, I initially opted for consistency and smaller diff.
But, done.

mattcaswell commented:
Also looks like there are some interesting travis failures being reported...mem leaks?

kaduk commented:
Yes, mem leaks, it looks like.

kaduk commented:
> I could believe that the 1.3 client is sending the new 1.3 ciphersuites and that's tripping up the hardcoded expectations.

Nope, looks like the client is not sending any 1.3-permitted ciphers but the server is still trying to pick TLS 1.3 (?!).  On the one hand, this is a rather silly set of ciphers for the client to send, but should the server really be picking TLS 1.3 when there are no supported ciphers?

davidben commented:
> should the server really be picking TLS 1.3 when there are no supported ciphers

Yes. I think it's reasonable to consider that a client misconfiguration and fail that. (Or rather you pick the version first and then fail to pick a cipher in the usual way.) In a future where clients and servers implement the [downgrade sentinel](https://tlswg.github.io/tls13-spec/#rfc.section.4.1.3), that kind of server fallback wouldn't work anyway. Also it adds complexity.

kaduk commented:
@davidben that seems to imply that openssl should not claim to support 1.3 as a client when it accepts a user-supplied cipher string that does not include 1.3-permitted ciphers.

davidben commented:
I suppose that's part of the TLS 1.3 cipher suite story that I think is still TBD for you all? On our end, not having to deal with this was a large motivator in deciding to decouple TLS 1.3 ciphers from the cipher string. We briefly tried saying that if you want TLS 1.3 you're responsible for making sure your other settings are consistent, which is normally true. But for cipher suites, that nearly immediately showed itself to be a deployment nuisance.

kaduk commented:
rebased yet again to pick up the no-ec fixes on master; on top of that, plug the memory leak and fix no-ec for these changes.

mattcaswell commented:
Conflicts again :-)

kaduk commented on ssl/ssl_lib.c:
```diff
@@ -4363,6 +4363,84 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
+}
+
+
 #endif
```
I bet this stuff shouldn't be inside the OPENSSL_NO_CT block...

kaduk commented:
rebased again; also moved tls_early_post_process_client_hello() to be next to the tls_process_client_hello() it was split from (as opposed to being next to the tls_post_process_client_hello() that it's now "part of"), to help git figure out what is going on -- it got confused because the code that Steve changed was considered "moved code" by the early callback change, whereas really it was the tls_early_post_process_client_hello() code that was moved, but git picks the shorter diff, for better or for worse.

kaduk commented:
@mattcaswell did you have other things to say than just noting the conflicts, or do you just need more time?  (I feel like you were focused on other things for the past couple days, and rightly so...)

kaduk commented:
rebased, 'make update', and squashed.  Now we need to conjure up a second reviewer ... should we ask @richsalz ?

richsalz commented:
Not me; someone from a different organization is better.

levitte commented:
From a different organization?  Isn't @mattcaswell good enough?

davidben commented on doc/man3/SSL_CTX_set_early_cb.pod:
```diff
@@ -0,0 +1,103 @@
+=head1 SYNOPSIS
+
+ typedef int (*SSL_early_cb_fn) (SSL *s, int *al, void *arg);
+ void SSL_CTX_set_early_cb(SSL_CTX *c, void *arg);
```
This looks like it's missing the `SSL_early_cb_fn` argument.

davidben commented on doc/man3/SSL_CTX_set_early_cb.pod:
```diff
@@ -0,0 +1,103 @@
+The application's supplied early callback returns 1 on success, 0 on failure,
+and a negative value to suspend processing.
+
+SSL_early_isv2() returns 1 for SSLv2-format ClientHellos and 0 otherwise.
```
Probably worth a sentence to clarify how such ClientHellos are presented by this API. (I.e. whether the random and cipher list are converted or exposed in their V2ClientHello form.)

richsalz commented:
By our intended policy, yes.  But I am going to try to avoid doing so when possible

levitte commented on test/handshake_helper.c:
```diff
@@ -123,6 +124,48 @@ static int select_server_ctx(SSL *s, void *arg, int ignore)
+        return 1;
+    }
+    return 0;
+}
```
This function worries me...  does it mean that someone who wants to implement these early callbacks must use internal functions?  That can't be good...

mattcaswell replied:
You mean the PACKET functions? No, you don't need to use them at all. @kaduk is just using them as a short cut

levitte replied:
Yeah, I think it's a bad idea for this...  it's better to test under conditions that our users will get into.

kaduk replied:
@levitte is 058cd485e3d4c0f7436109e46f96780c2a32ad36 enough to resolve your "requested changes"?

levitte replied:
Yes, this does resolve the issue

levitte commented on test/build.info:
```diff
@@ -260,7 +260,7 @@ IF[{- !$disabled{tests} -}]
   INCLUDE[testutil.o]=..
   INCLUDE[ssl_test_ctx.o]=../include
-  INCLUDE[handshake_helper.o]=../include
+  INCLUDE[handshake_helper.o]=../include ../ssl
```
Unless the intended program is meant to test internals, this isn't a good idea.  Does this even link correctly with shared libraries?

kaduk replied:
I'm pretty sure I only pulled in -I../ssl for packet_locl.h, since using PACKET_* is a couple orders of magnitude nicer than not using them.  But, I guess it is a slippery slope to have that sitting there, in that something less innocuous could more easily slip in.

levitte replied:
Ah, I forget that the PACKET API is all inline.  But still, yeah, slippery slope 'n all that.

kaduk replied:
Okay, pushed an update that replaces the PACKET usage with unrolled accesses (that are twice as long).

levitte commented:
I'm seeing a test issue in the builds.  The SNI test is failing...

kaduk commented:
> I'm seeing a test issue in the builds. The SNI test is failing...

Okay, I'll rebase and take a closer look.

kaduk commented:
Post-rebase, I can't reproduce the SNI test failure (I think we were running into the issue fixed by e4a3d0f96800e578ddb76dcc387125df65641500), but a local --strict-warnings build did find a couple of issues with the last commit, that are fixed and squashed.
(It turns out that there are no sequence points in the RHS of "len = (*(p++) << 1) + *(p++);", so the presence of two p++ is potentially undefined behavior.)

kaduk commented:
Hmm, just a doc-nits issue from travis, apparently.

richsalz commented:
you must fix all doc nits you know :)

levitte commented on doc/man3/SSL_CTX_set_early_cb.pod:
```diff
@@ -0,0 +1,110 @@
+
+=head1 SYNOPSIS
+
+ typedef int (*SSL_early_cb_fn) (SSL *s, int *al, void *arg);
```
I think that space between ) and ( is the cause of one of the doc-nits issues...

levitte commented on doc/man3/SSL_CTX_set_early_cb.pod:
```diff
@@ -0,0 +1,110 @@
+
+=head1 NAME
+
+SSL_CTX_set_early_cb, SSL_early_isv2, SSL_early_get0_legacy_version, SSL_early_get0_random, SSL_early_get0_session, SSL_early_get0_ciphers, SSL_early_get0_compression_methods, SSL_early_get0_ext - callback functions for early server-side ClientHello processing
```
`SSL_early_get0_session` should be `SSL_early_get0_session_id`...

kaduk commented:
Thanks for the pointer about the space in ') (', @levitte , that would have been rather nonobvious from the given error message.
I went with the [skip ci] route rather than push -f, since the CI seems to have quite a backlog.

kaduk commented:
Who should merge it?

richsalz commented:
Per standard practice, it will usually be one of the reviewers.  It must be someone on the team of course :)

levitte commented:
I can do it.

levitte commented:
Merged!  fc5ece2ee4c7ef41759c2267eccbe08df3245721 ..  2afaee5193c3f567ea00ad613f3ee82f985b7141

levitte commented:
(I did some editorial work...  the last commits were obviously fixups, so I squashed them with the appropriate original commits)
